src/prepare.rs

```diff
@@ -3,10 +3,75 @@
 use crate::*;
 pub use std::fmt::Write;
 
-#[derive(Debug, Default)]
-pub struct SqlWriter {
-    pub(crate) counter: usize,
-    pub(crate) string: String,
+pub trait SqlWriter: Write + ToString {
+    fn push_param(&mut self, value: Value, query_builder: &dyn QueryBuilder);
+
+    fn as_writer(&mut self) -> &mut dyn Write;
+}
+
+impl SqlWriter for String {
+    fn push_param(&mut self, value: Value, query_builder: &dyn QueryBuilder) {
+        self.push_str(&query_builder.value_to_string(&value))
+    }
+
+    fn as_writer(&mut self) -> &mut dyn Write {
+        self as _
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SqlWriterValues {
+    counter: usize,
+    placeholder: String,
+    numbered: bool,
+    string: String,
+    values: Vec<Value>,
+}
+
+impl SqlWriterValues {
+    pub fn new(placeholder: &str, numbered: bool) -> Self {
+        let placeholder = placeholder.to_owned();
+        Self {
+            counter: 0,
+            placeholder,
+            numbered,
+            string: String::with_capacity(256),
+            values: Vec::new(),
+        }
+    }
+
+    pub fn into_parts(self) -> (String, Values) {
+        (self.string, Values(self.values))
+    }
+}
+
+impl Write for SqlWriterValues {
+    fn write_str(&mut self, s: &str) -> std::fmt::Result {
+        write!(self.string, "{}", s)
+    }
+}
+
+impl ToString for SqlWriterValues {
+    fn to_string(&self) -> String {
+        self.string.clone()
+    }
+}
```

Comment:
Is this necessary? I think it'd be better if to_string is the version with parameters injected rather than just a clone of the SQL with placeholder

Reply:
My thoughts were: injection parameters are expensive. If someone needs an inject parameter into sql query it is better to call it explicit. 
Do I need to fix that?

Reply:
Umm I think we have two options 1) remove this impl 2) make it also inject parameters. I am okay with either.

Reply:
@tyt2y3 I have problems:

1. I cannot remove this impl, because I need it, for example: https://github.com/SeaQL/sea-query/blob/master/src/query/traits.rs#L21
2. I cannot make `inject_parameters` because I need: `query_builder` but I didnot store it in SqlWriter...